torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1479,38 +1478,46 @@ def _reset_lazy_init(self) -> None:
         self._init_reshard_after_forward()
 
     def _lazy_init(self) -> None:
-        """Initialization steps that should happen lazily, typically right
-        before the first forward pass.
         """
-        # Initialize param attributes lazily, in case the param's dtype or
-        # device changes after __init__.
-        for p in self.params:
-            self._init_param_attributes(p)
-
-        # Initialize _is_root and setup streams. These steps would ideally
-        # happen in __init__, but _is_root can only be determined after the
-        # entire model hierarchy is setup, thus we run it lazily.
-        if self._is_root is None:
-            # _is_root means that we are in the outermost module's forward.
-            self._set_is_root()
-            self._setup_streams()
-
-        if self._is_root:
-            # Buffers stay on GPU, and don't get sharded. Since _cast_buffers
-            # applies recursively, we only call this from the root instance.
-            self._cast_buffers(recurse=True)
-
-            # Don't free the full params for the outer-most (root) instance,
-            # In most cases, root instance contains params in the last layers
-            # or has no params. In these cases, those params will be needed
-            # immediately after for the backward pass. Note that this only
-            # applies currently when freeing parameters at end of layer's
-            # forward pass.
-            self.reshard_after_forward = False
+        Performs initialization lazily, typically right before the first
+        forward pass. The laziness is needed to ensure that the parameter
+        device/dtype and the FSDP hierarchy have finalized.
 
-            # Due to the use of streams, we need to make sure the previous
-            # ``optim.step()`` is done before we all-gather parameters.
-            self._wait_for_previous_optim_step()
+        This method's actual logic only runs on the root FSDP instance, which
+        performs initialization for all non-root FSDP instances to avoid
+        partial initialization.
+        """
```

Comment:
There were two components to the lazy initialization:
1. Setting each FSDP instance's `_is_root` and sharing some data structures (streams, execution order info, graph order) from the root instance to non-root instances.
2. Initializing parameter attributes (via `_init_param_attributes()`).

For 1, the root FSDP instance does it all at once, meaning it is done when the root FSDP instance's `_lazy_init()` returns. In contrast, for 2, each FSDP instance is responsible for itself.

This exposes a window between the end of the root's `_lazy_init()` and a non-root's `_lazy_init()`, where the non-root is partially initialized. There is no performance downside to this, but logically, I think it is clearer if the initialization is done atomically with respect to one `_lazy_init()` call.

This means that `self._is_root is None` fully indicates whether an FSDP instance's initialization has happened or not. Before, for a non-root FSDP instance, it could be that `self._is_root == False`, but `_init_param_attributes()` had not been called yet.

This also means that we have a clear separation between the logic that is expected to run every iteration (e.g. `_wait_for_previous_optim_step()`) versus the one-time initialization logic.

Reply:
The early return in `_init_param_attributes()` can actually be removed now since it is only ever called once, but since there is an `assert` for CPU offloading, I decided to leave it in.